include/zephyr/bluetooth/encrypted_ad_data.h

```diff
@@ -0,0 +1,77 @@
+ */
+#define BT_EAD_RANDOMIZER_DIRECTION_BIT 7
+
+/* Fixed value used for the Additional Authenticated Data (see Supplement to the Bluetooth Core
+ * Specification v11, Part A 1.23.3)
+ */
+static uint8_t bt_ead_aad[] = {0xEA};
+BUILD_ASSERT(sizeof(bt_ead_aad) == BT_EAD_AAD_SIZE);
+
+/**
+ * @brief Encrypt and authenticate the given advertising data.
+ * The resulting data in \p encrypted_ad_data will look like that:
+ *     - Randomizer is added in the 5 first bytes;
+ *     - AAD is added in the next byte;
+ *     - Encrypted payload is added (payload_size bytes);
+ *     - MIC is added in the last 4 bytes.
+ *
+ * @note The function must be called each time the RPA is updated or the data
+ * are modified.
+ *
+ * @note The term `advertising strucure` is used to describe the advertising
+ *       data with the advertising type and the length of those two.
+ *
+ * @param[in]  session_key Key of 16 bytes used for the encryption.
+ * @param[in]  iv Initialisation Vector used to generate the nonce. It must be
+ *             changed each time the Session Key change.
+ * @param[in]  payload Advertising Data to encrypt. Can be multiple advertising
+ *             structure concatenated.
+ * @param[in]  payload_size Size of the Advertising Data to encrypt.
+ * @param[out] encrypted_ad_data Encrypted Adv Data including the Randomizer and
+ *             the MIC. Size must be `BT_EAD_RANDOMIZER_SIZE + BT_EAD_AAD_SIZE +
+ *             payload_size + BT_EAD_MIC_SIZE`.
+ * @return int Size of \p encrypted_ad_data in bytes if successful, else return
+ *             `-EIO` if an error occurred during the encryption.
+ */
+int bt_ead_encrypt_ad_data(const uint8_t *session_key, const uint8_t *iv, const uint8_t *payload,
+			   size_t payload_size, uint8_t *encrypted_ad_data);
+
+/**
+ * @brief Decrypt and authenticate the given encrypted advertising data.
+ *
+ * @note The term `advertising strucure` is used to describe the advertising
+ *       data with the advertising type and the length of those two.
+ *
+ * @param[in]  session_key Key of 16 bytes used for the encryption.
+ * @param[in]  iv Initialisation Vector used to generate the `nonce`.
+ * @param[in]  encrypted_payload Encrypted Advertising Data received. This
+ *             should only contain the advertising data and not the length nor
+ *             the type.
+ * @param[in]  encrypted_ad_data_size Size of the received advertising data in
+ *             bytes. Should be equal to the length minus the size of the type (1 byte).
+ * @param[out] payload Decrypted advertising structure. Can contain multiple
+	       advertising structure.
+ * @return int Size of \p payload in bytes if the decryption is successful, else
+ *             return `-EIO` if an error occurred during the decryption.
+ */
+int bt_ead_decrypt_ad_data(const uint8_t *session_key, const uint8_t *iv,
+			   const uint8_t *encrypted_payload, size_t encrypted_ad_data_size,
+			   uint8_t *payload);
```

Comment:
What you propose is dangerous. How will the application know if an AD is authenticated or not? A simple API is much easier to reason about for our users, and that benefits security. The intention is to have a companion library in the near future that handles the complexity that is in the test code in this PR. We want to define a simple API that can be built on top of. An API that is general because it's simple is better than an API that is general because it has complex extensions.

Reply:
See https://github.com/zephyrproject-rtos/zephyr/pull/55225#discussion_r1119800205

> How will the application know if an AD is authenticated or not?

What is the use case for this? Why should encrypted advertising be considered differently than e.g. encrypted ACL data?

Reply:
> See [#55225 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/55225#discussion_r1119800205)
> 
> > How will the application know if an AD is authenticated or not?
> 
> What is the use case for this? Why should encrypted advertising be considered differently than e.g. encrypted ACL data?

Messages are authenticated to establish trust. Use case: An electronic shelf label must not accept updates that are unauthenticated, as that would open up for a trivial attack of just spoofing an advertisement with a plain-text unauthenticated update.